org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/services/common/file/RemoteFileService.java

```diff
@@ -256,43 +257,57 @@ private File getFileInternal(RemoteFile spec) {
     if (canonicalSimpleName == null || !canonicalSimpleName.equals(spec.getName().startsWith("/") ? spec.getName().substring(1) : spec.getName())) {
       throw new SecurityException("invalid or unaccessible path");
     }
-    //
-    String filename = canonicalSimpleName;
-    if (spec.getLocale() != null && filename.lastIndexOf('.') != -1) {
-      //check locale string for hacking patterns (only allow
-      String localeText = spec.getLocale().toString().replaceAll("__", "_");
-      if (!LOCALE_SECURITY_PATTERN.matcher(localeText).matches()) {
-        throw new SecurityException("invalid or unaccessible path");
-      }
-      String[] checkedLocaleParts = localeText.split("_", 3);
-      String prefix = filename.substring(0, filename.lastIndexOf('.'));
-      String suffix = filename.substring(filename.lastIndexOf('.'));
-      for (int i = checkedLocaleParts.length - 1; i >= 0; i--) {
-        if (prefix.toLowerCase().endsWith(checkedLocaleParts[i].toLowerCase())) {
-          prefix = prefix.substring(0, prefix.length() - checkedLocaleParts[i].length());
-          if (prefix.endsWith("_")) {
-            prefix = prefix.substring(0, prefix.length() - 1);
-          }
-        }
-      }
-      if (!prefix.endsWith("_")) {
-        prefix = prefix + "_";
+    String filename = tryAddLocaleToFileName(canonicalSimpleName, spec.getLocale(), canonicalFolder, file -> file.exists());
+    File file = new File(canonicalFolder, filename);
+    return file;
+  }
+
+  private static final int LOCALE_TEXT_MAX_LENGTH = 64;
+  private static final Pattern LOCALE_SECURITY_PATTERN = Pattern.compile("[-a-z0-9_.#]+", Pattern.CASE_INSENSITIVE);
+
+  protected String tryAddLocaleToFileName(String filename, Locale locale, String canonicalFolder, Predicate<File> fileExistsPredicate) {
+    int suffixSeparatorIndex = filename.lastIndexOf('.');
+    if (locale == null || suffixSeparatorIndex == -1) {
+      return filename;
+    }
+    String localeText = locale.toString().replaceAll("__", "_");
+    if (localeText.length() > LOCALE_TEXT_MAX_LENGTH) {
+      return filename;
+    }
+    // make sure only valid locale characters are accepted.
+    if (!LOCALE_SECURITY_PATTERN.matcher(localeText).matches()) {
+      throw new SecurityException("invalid or inaccessible path");
+    }
+
+    String[] checkedLocaleParts = Arrays.stream(localeText.split("_", 3))
```

Comment:
You could use Pattern.compile(..).splitAsStream() here? Maybe you can also save the pattern in a constant? The limit can be implemented using limit()